src/extra/components/detector_motors.py

```diff
@@ -0,0 +1,414 @@
+                If any motor position key is not found
+        """
+        self.detector_id = detector_id
+        self.src_ptrn = src_ptrn
+        self.key_ptrn = key_ptrn
+        self.dims = dims
+
+        names, coordinates = zip(*self.dims.items())
+        self.motor_ids = [(
+                src_ptrn.format(**dict(zip(names, values))),
+                key_ptrn.format(**dict(zip(names, values)))
+            ) for values in product(*coordinates)]
+        self.motor_labels = [
+            ''.join(f"{n.upper()}{v}" for n, v in zip(names, values))
+            for values in product(*coordinates)]
+
+        self.shape = tuple(len(c) for c in coordinates)
+        self.num_sources = np.prod(self.shape)
+        self.dc = dc.select(self.motor_ids)
+
+        self.num_trains = len(self.dc.train_ids)
+
+        # check sources
+        self.keys = []
+        for src, key in self.motor_ids:
+            self.keys.append(self.dc[src, key])
+
+        # add train_ids
+        self.train_ids = np.arange(
+            self.dc.train_ids[0], self.dc.train_ids[-1] + np.uint64(1),
+            dtype=np.uint64)
+
+    def positions(self, labelled=True):
+        """Returns the motor positions for all trains.
+
+        Args:
+            labelled (bool):
+                If True, returns the xarray with labelled dimensions,
+                overwise returns numpy.ndarray
+
+        Returns:
+            positions (numpy.ndarray or xarray.DataArray):
+                The motor positions
+        """
+        train_ids, pos, counts = self._read_positions()
+        pos = np.repeat(pos, counts, axis=0)
+
+        if labelled:
+            dims = ["trainId"] + list(self.dims.keys())
+            coords = {"trainId": self.train_ids}
+            coords.update({name: values for name, values in self.dims.items()})
+            return xarray.DataArray(
+                pos, dims=dims, coords=coords, name=self._position_key)
+        else:
+            return pos
+
+    def _read_positions(self):
+        """Reads and compresses motor positions."""
+        if hasattr(self, "_positions"):
+            return self._positions
+
+        # read positions
+        pos = np.zeros((self.num_trains, self.num_sources), dtype=float)
+        for source_no, key_data in enumerate(self.keys):
+            pos[:, source_no] = key_data.ndarray()
+        pos = pos.reshape(-1, *self.shape)
+
+        # compress
+        axes = tuple(range(1, pos.ndim))
+        ix = np.flatnonzero(
+            np.insert(np.any(np.diff(pos, axis=0), axis=axes), 0, True))
+
+        # train ids when positions are changed
+        train_ids = np.asarray(self.dc.train_ids)[ix]
+        # count of trains at every position
+        counts = np.diff(np.append(train_ids.astype(int),
+                                   int(self.dc.train_ids[-1]) + 1))
+
+        self._positions = train_ids, pos[ix], counts
+        return self._positions
+
+    def compressed_positions(self, labelled=True):
```

Comment:
This is rather nitpicky, but I don't really like the generic term `labelled`. Both the EXtra components and extra-data functions use that name to return either `Series` or `DataArray` objects, which feels confusing/inconsistent to me :shrug:

I'd prefer a more explicit `xarray` argument, though I know this would add even more inconsistency so I don't insist on it.

Reply:
I agree, but I tend to follow the extra-data naming until we agree another convention.